liquidator/src/amm/AstroportPoolProvider.ts

```diff
@@ -0,0 +1,231 @@
+        }
+
+        const poolQueries = this.pairs.map((pair) => this.producePoolQuery(pair) )
+        
+
+        // execute the query
+        while (retries < this.maxRetries) {
+            try {
+                const response = await fetch(this.graphqlEndpoint, {
+                    method: 'post',
+                    headers: {
+                        'Content-Type': 'application/json',
+                    },
+                    body: JSON.stringify(poolQueries),
+                });
+         
+                const responseData : PoolResponseData[] =await response.json()
+        
+                return responseData.map((poolResponse, index) => {
+        
+                    // Our address is our key. We add this in to our graphql queries so we can identify the correct pool
+                    const address : string = Object.keys(poolResponse.data)[0]
+        
+                    const queryData : ContractQueryPool = poolResponse.data[address]!.contractQuery
+                    const pool : Pool = {
+                        address : address,
+                        id : index as unknown as Long,
+                        poolAssets : this.producePoolAssets(queryData.assets),
+                        swapFee : "0.00",
+                    }
+        
+                    return pool
+                })
+            } catch(err) {
+                console.error(err)
+                retries += 1
+                await sleep(1000)
+            }
+        }
+
+        return []
+    }
+
+    async fetchPairContracts(contractAddress: string, limit = 10): Promise<Pair[]> {
+        let startAfter = this.findLatestPair(this.pairs)
+        let retries = 0
+
+        const pairs : Pair[] = []
+
+        // Loop until we find all the assets or we hit our max retries
+        while (retries < this.maxRetries) {
+            try {
+            
+            const query = this.producePairQuery(startAfter, limit, contractAddress)
+
+            const response = await fetch(this.graphqlEndpoint, {
+                method: 'post',
+                headers: {
+                    'Content-Type': 'application/json',
+                },
+                body: JSON.stringify(query),
+            });
+        
+            // get our data
+            const responseData : ResponseData = await response.json();
+            const contractQueryData = responseData.data.wasm.contractQuery as ContractQueryPairs
+
+            // Filter out pairs that are not XYK
+            const batchPairs = contractQueryData.pairs
+                                .filter((pair) => pair.pair_type.xyk !== undefined)
+            
+            if (batchPairs.length === 0) {
+                // we have no more pairs to load
+                return pairs
+            } else {
+
+                batchPairs.forEach((pair) => {
+                    pairs.push(pair)
+                })
```

Comment:
```suggestion
                pairs.concat(batchPairs)
```
I think this might be equivalent